pkg/log/context.go

```diff
@@ -0,0 +1,184 @@
+	FieldComponentName = "component"
+	// FieldCorrelationID is the key of the correlation id field in the log message.
+	FieldCorrelationID = "correlation_id"
+)
+
+// Contexter interface
+type Contexter interface {
+	Context() context.Context
+}
+
+// Settings type to be loaded from the environment
+type Settings struct {
+	Level  string
+	Format string
+}
+
+// DefaultSettings returns default values for Log settings
+func DefaultSettings() *Settings {
+	return &Settings{
+		Level:  "debug",
+		Format: "text",
+	}
+}
+
+// Validate validates the logging settings
+func (s *Settings) Validate() error {
+	if len(s.Level) == 0 {
+		return fmt.Errorf("validate Settings: LogLevel missing")
+	}
+	if len(s.Format) == 0 {
+		return fmt.Errorf("validate Settings: LogFormat missing")
+	}
+	return nil
+}
+
+// Configure creates a new context with a logger using the provided settings.
+func Configure(ctx context.Context, settings *Settings) context.Context {
+	once.Do(func() {
+		level, err := logrus.ParseLevel(settings.Level)
+		if err != nil {
+			panic(fmt.Sprintf("Could not parse log level configuration: %s", err))
+		}
+		formatter, ok := supportedFormatters[settings.Format]
+		if !ok {
+			panic(fmt.Sprintf("Invalid log format: %s", settings.Format))
+		}
+		logrus.SetLevel(level)
+		logrus.SetFormatter(formatter)
+		logger := &logrus.Logger{
+			Formatter: formatter,
+			Level:     level,
+			Out:       os.Stdout,
+			Hooks:     make(logrus.LevelHooks),
+		}
+		defaultEntry = logrus.NewEntry(logger)
+	})
+	return ContextWithLogger(ctx, defaultEntry)
+}
+
+// ForContext retrieves the current logger from the context, configured for the provided component.
+// Optionally keys mapped to values from the context can be provided.
+// If no logger is present in the context, the default logger is returned
+func ForContext(ctx context.Context, component string, keys ...interface{}) *logrus.Entry {
+	entry := ctx.Value(logKey{})
+	if entry == nil {
+		// copy so that changes to the new entry do not reflect the default entry
+		entry = copyEntry(defaultEntry)
+	}
+	fields := make(logrus.Fields, len(keys)+1)
+	fields[FieldComponentName] = component
+	for _, key := range keys {
+		value := ctx.Value(key)
+		if value != nil {
+			fields[fmt.Sprint(key)] = value
+		}
+	}
+	logEntry := entry.(*logrus.Entry).WithFields(fields)
+	if _, exists := logEntry.Data[FieldCorrelationID]; !exists {
+		logEntry = logEntry.WithField(FieldCorrelationID, "default")
+	}
+	contextLogLevel, exists := ctx.Value(LevelKey{}).(string)
+	if exists {
+		level, err := logrus.ParseLevel(contextLogLevel)
+		if err != nil {
+			logEntry.Warnf("Dynamic log level change not supported for log level %s", contextLogLevel)
+		} else {
+			logEntry.Logger.Level = level
+		}
+	}
+	return logEntry
+}
+
+// ForContextProvider retrieves the current logger from the context provided.
+func ForContextProvider(contexter Contexter, component string, keys ...interface{}) *logrus.Entry {
```

Comment:
would it make sense to drop this one and have only C and D/

Reply:
it's an option but you'd have request.Context() everywhere then
or in every handler you'd have 1 line `ctx := request.Context()` and then `log.C(ctx)`